Gems/LmbrCentral/Code/Tests/TubeShapeTest.cpp

```diff
@@ -412,4 +413,37 @@ namespace UnitTest
             EXPECT_THAT(variableRadius, FloatEq(radiis.second));
         }
     }
-}
+
+    TEST_F(TubeShapeTest, ShapeHasThreadsafeGetSetCalls)
+    {
+        // Verify that setting values from one thread and querying values from multiple other threads in parallel produces
+        // correct, consistent results.
+
+        // Create our sphere centered at 0 with half our height as the radius.
+        AZ::Entity entity;
+        CreateTube(AZ::Transform::CreateTranslation(AZ::Vector3::CreateZero()), ShapeThreadsafeTest::ShapeHeight / 2.0f, entity);
+
+        // Define the function for setting unimportant dimensions on the shape while queries take place.
+        auto setDimensionFn = [](AZ::EntityId shapeEntityId, float minDimension, uint32_t dimensionVariance, float height)
+        {
+            // Set the radius back to the same value. This should have no effect.
+            LmbrCentral::TubeShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::TubeShapeComponentRequestsBus::Events::SetRadius, height / 2.0f);
+
+            // Set the end radii to random values. As long as we don't set the middle radii that describe the portion of the tube
+            // that intersects the origin, changing these should have no effect on our test point distance.
+
+            float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            LmbrCentral::TubeShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::TubeShapeComponentRequestsBus::Events::SetVariableRadius, 0, radius);
+
+            radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
+            LmbrCentral::TubeShapeComponentRequestsBus::Event(
+                shapeEntityId, &LmbrCentral::TubeShapeComponentRequestsBus::Events::SetVariableRadius, 3, radius);
```

Comment:
You can pass in another lambda to the TubeShapeComponentRequestsBus so that the only 1 EBus event occurs instead of 3. That would improve performance of the test.
```
auto SetRadii = [](LmbrCentral::TubeSharedComponentRequests* requests)
{
    requests->SetRadius(height / 2.0f)
    float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
    requests->SetVariableRadius, 0, radius);
    radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
    requests->SetVariableRadius, 0, radius);
};

LmbrCentral::TubeShapeComponentRequestsBus::Event(shapeEntityId, SetRadii);
```

Reply:
I'd rather leave this one as separate EBus calls, as it will do a better job of spamming the EBus mutex with smaller lock/unlock cycles. 

Also, FWIW, this change makes the test performance worse - across 10 runs the average time went from 64.1 ms to 79.5 ms. The way the test works is that it runs the _query_ threads for a given number of iterations, and runs the Set thread as many times as necessary until the queries are done. By having the Set thread hold a single lock for longer, all 4 query threads are slowed down by a small amount on each iteration.

Reply:
The performance shouldn't be worse.
Was this logic added inside of the `SetDimensionFn` function?

Reply:
Yes. Here is the specific test that I measured vs what is in the PR:
```
    TEST_F(TubeShapeTest, ShapeHasThreadsafeGetSetCalls)
    {
        // Verify that setting values from one thread and querying values from multiple other threads in parallel produces
        // correct, consistent results.

        // Create our sphere centered at 0 with half our height as the radius.
        AZ::Entity entity;
        CreateTube(AZ::Transform::CreateTranslation(AZ::Vector3::CreateZero()), ShapeThreadsafeTest::ShapeHeight / 2.0f, entity);

        // Define the function for setting unimportant dimensions on the shape while queries take place.
        auto setDimensionFn = [](AZ::EntityId shapeEntityId, float minDimension, uint32_t dimensionVariance, float height)
        {
            auto SetRadii = [minDimension, dimensionVariance, height](LmbrCentral::TubeShapeComponentRequests* requests)
            {
                requests->SetRadius(height / 2.0f);
                float radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
                requests->SetVariableRadius(0, radius);
                radius = minDimension + aznumeric_cast<float>(rand() % dimensionVariance);
                requests->SetVariableRadius(3, radius);
            };

            LmbrCentral::TubeShapeComponentRequestsBus::Event(shapeEntityId, SetRadii);
        };

        // Run the test, which will run multiple queries in parallel with each other and with the dimension-setting function.
        // The number of iterations is arbitrary - it's set high enough to catch most failures, but low enough to keep the test
        // time to a minimum.
        const int numIterations = 30000;
        ShapeThreadsafeTest::TestShapeGetSetCallsAreThreadsafe(entity, numIterations, setDimensionFn);
    }
```
I ran this with a gtest_repeat=10. The new averages with 30000 iterations are 28.4 ms for what's in the PR vs 31.4 ms for this version.

There isn't an easy way to capture a picture of what's happening here (I think I'd have to add the Profiler Gem to the LmbrCentral tests to show this in PIX), but my assumption for why it's slower is that the Set thread is successfully stealing more time from the query threads, making them take longer overall to finish their given number of iterations.

Basically, we've taken each loop of the Set thread from looking like this:
|----| |---| |----|
to this:
|-----------|

There are less "gaps" where the query threads have an ability to acquire their mutex, so the Set thread can potentially block all 4 query threads for longer stretches of time on each of its iterations. 

I admit I don't fully understand the threading nuances going on here though. Given that picture, I would've expected to find that the Set thread was iterating _more_ frequently with the above code change and starving out the query threads, but in fact it's executing _less_ iterations. The Set loop is executing on average 8034 iterations in the PR code, vs 7029 iterations in the code above. So it's running less times, but it must also manage to block the query threads for longer when it does run, because the query thread execution length is what dictates the total runtime length. 

I agree that this is a non-intuitive result, and in _general_ I believe in grouping EBus calls together like this, but for this particular test it doesn't appear to be a good idea.